File: 6-zigzag-conversion/6-zigzag-conversion.java
class Solution {
    public String convert(String s, int numRows) {
        String res = "";
        int n = numRows;
        if(n == 1) return s;
        for(int i = 0; i < numRows; i++) {
            if(i == 0 || i == numRows - 1) {
                for(int j = i; j < s.length(); j += 2* n - 2) {
                    res += s.charAt(j);
                }
            } else {
                for(int j = i, k = 2* n - 2 - i; j < s.length() || k < s.length(); j+= 2*n - 2, k += 2*n - 2) {
                    if(j < s.length()) res += s.charAt(j);
                    if(k < s.length()) res += s.charAt(k);
                }
            }
        }
        return res;
    }
}

File: 6-zigzag-conversion/6-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        res = ""
        if len(s) == 1 or numRows == 1: return s
        for i in range(numRows):
            if i == 0  or i == numRows - 1:
                for j in range(i, len(s), 2*(numRows) - 2):
                    res += s[j]
            else:
                j , k = i, 2 *(numRows) - 2 - i
                while j < len(s) or k < len(s):
                    res += s[j]
                    j += 2*(numRows) - 2
                    if k < len(s):
                        res += s[k]
                        k += 2 * numRows - 2
        return res